File: Object/Surface.cpp
#include "Object/Surface.hpp"

#include "Object/Albedo.hpp"
#include "Object/Brdf.hpp"
#include "Object/Intersection.hpp"

#include <cmath>

namespace Object {
    Surface::Surface(std::unique_ptr<Object::Albedo> albedo, std::vector<std::unique_ptr<Object::Brdf>> brdfs, float transmitIor, const Math::Radiance &radiance, std::unique_ptr<Object::NormalMap> normalMap)
    {
        mAlbedo = std::move(albedo);
        mBrdfs = std::move(brdfs);
        mTransmitIor = transmitIor;
        mRadiance = radiance;
        mNormalMap = std::move(normalMap);

        mLambert = 0;
        mOpaque = false;
        for(const std::unique_ptr<Object::Brdf> &brdf : mBrdfs) {
            if(brdf->opaque()) {
                mOpaque = true;
            }

            if(brdf->lambert() > 0) {
                mLambert = brdf->lambert();
            }
        }

        mTransmitIor = transmitIor;        
    }

    const Object::Albedo &Surface::albedo() const
    {
        return *mAlbedo;
    }

    const Math::Radiance &Surface::radiance() const
    {
        return mRadiance;
    }

    bool Surface::hasNormalMap() const
    {
        return mNormalMap.get();
    }

    const Object::NormalMap &Surface::normalMap() const
    {
        return *mNormalMap;
    }

    Math::Color Surface::reflected(const Object::Intersection &isect, const Math::Vector &dirIn) const
    {
        Math::Color col;
        Math::Color colTransmit(1, 1, 1);

        for(const auto &brdf : mBrdfs) {
            col = col + colTransmit * brdf->reflected(dirIn, isect.facingNormal(), -isect.ray().direction(), isect.albedo());
            colTransmit = colTransmit * brdf->transmitted(dirIn, isect.facingNormal(), isect.albedo());
        }

        return col;
    }

    Math::Color Surface::transmitted(const Object::Intersection &isect, const Math::Vector &dirIn) const
    {
        Math::Color colTransmit(1, 1, 1);

        for(const auto &brdf : mBrdfs) {
            colTransmit = colTransmit * brdf->transmitted(dirIn, -isect.facingNormal(), isect.albedo());
        }

        return colTransmit;
    }

    Math::Color Surface::sample(const Object::Intersection &isect, Math::Sampler &sampler, Math::Vector &dirIn, Math::Pdf &pdf) const
    {
        const Math::Vector dirOut = -isect.ray().direction();
        const Math::Normal &nrmFacing = isect.facingNormal();

        float transmitThreshold = 0;
        if(!opaque()) {
            bool reverse = (isect.normal() * dirOut < 0);

            float ratio = 1.0f / mTransmitIor;
            if (reverse) {
                ratio = 1.0f / ratio;
            }

            float c1 = dirOut * nrmFacing;
            float c2 = std::sqrt(1.0f - ratio * ratio * (1.0f - c1 * c1));

            dirIn = Math::Vector(nrmFacing) * (ratio * c1 - c2) - dirOut * ratio;
            Math::Color throughput = transmitted(isect, -dirOut);
            transmitThreshold = std::min(1.0f, throughput.maximum());
            float roulette = sampler.getValue();

            if(roulette < transmitThreshold) {
                pdf = Math::Pdf(1.0f, true);

                return transmitted(isect, dirIn) / (dirOut * nrmFacing * transmitThreshold);
            }
        }

        int idx = 0;
        if(mBrdfs.size() > 1) {
            float sample = sampler.getValue();
            idx = std::min((int)std::floor(mBrdfs.size() * sample), (int)mBrdfs.size() - 1);
        }
        const Object::Brdf &brdf = *mBrdfs[idx];
       
        dirIn = brdf.sample(sampler, nrmFacing, dirOut);
        pdf = Surface::pdf(isect, dirIn);
        return reflected(isect, dirIn) / (1 - transmitThreshold);
    }

    Math::Pdf Surface::pdf(const Object::Intersection &isect, const Math::Vector &dirIn) const
    {
        const Math::Vector dirOut = -isect.ray().direction();
        const Math::Normal &nrmFacing = isect.facingNormal();
        
        float totalPdf = 0;
        for(const std::unique_ptr<Object::Brdf> &brdf : mBrdfs) {
            totalPdf += brdf->pdf(dirIn, nrmFacing, dirOut);
        }
        totalPdf /= (float)mBrdfs.size();

        return totalPdf;
    }

    bool Surface::opaque() const
    {
        return mOpaque;
    }

    float Surface::lambert() const
    {
        return mLambert;
    }

    float Surface::transmitIor() const
    {
        return mTransmitIor;
    }

    void Surface::writeProxy(SurfaceProxy &proxy, OpenCL::Allocator &clAllocator) const
    {
        mRadiance.writeProxy(proxy.radiance);
        mAlbedo->writeProxy(proxy.albedo, clAllocator);
        proxy.numBrdfs = mBrdfs.size();
        proxy.brdfs = clAllocator.allocateArray<BrdfProxy>(proxy.numBrdfs);
        for(int i=0; i<mBrdfs.size(); i++) {
            mBrdfs[i]->writeProxy(proxy.brdfs[i]);
        }
        proxy.opaque = mOpaque;
        proxy.transmitIor = mTransmitIor;
        if(mNormalMap) {
            proxy.normalMap = clAllocator.allocate<NormalMapProxy>();
            mNormalMap->writeProxy(*proxy.normalMap, clAllocator);
        } else {
            proxy.normalMap = nullptr;
        }
    }

}

File: src/object/surface.rs
use crate::geo;
use crate::object;

use geo::Vec3;

use object::Albedo;
use object::Brdf;
use object::Color;
use object::Intersection;
use object::NormalMap;
use object::Radiance;
use object::Sampler;

pub struct Surface {
    pub albedo: Box<dyn Albedo>,
    pub brdfs: Vec<Box<dyn Brdf>>,
    pub transmit_ior: f32,
    pub radiance: Radiance,
    pub normal_map: Option<NormalMap>,
    pub opaque: bool,
}

impl Surface {
    pub fn new(
        albedo: Box<dyn Albedo>,
        brdfs: Vec<Box<dyn Brdf>>,
        transmit_ior: f32,
        radiance: Radiance,
        normal_map: Option<NormalMap>,
    ) -> Surface {
        let mut opaque = false;
        for brdf in brdfs.iter() {
            if brdf.opaque() {
                opaque = true;
            }
        }

        Surface {
            albedo,
            brdfs,
            transmit_ior,
            radiance,
            normal_map,
            opaque,
        }
    }

    pub fn reflected(&self, isect: &Intersection, dir_in: Vec3) -> Color {
        let mut color = Color::ZERO;
        let mut color_transmit = Color::ONE;

        for brdf in self.brdfs.iter() {
            color += color_transmit
                * brdf.reflected(
                    dir_in,
                    isect.facing_normal,
                    -isect.ray.direction,
                    isect.albedo,
                );
            color_transmit =
                color_transmit * brdf.transmitted(dir_in, isect.facing_normal, isect.albedo);
        }

        return color;
    }

    pub fn transmitted(&self, isect: &Intersection, dir_in: Vec3) -> Color {
        let mut color_transmit = Color::ONE;

        for brdf in self.brdfs.iter() {
            color_transmit =
                color_transmit * brdf.transmitted(dir_in, -isect.facing_normal, isect.albedo);
        }

        return color_transmit;
    }

    pub fn sample(
        &self,
        isect: &Intersection,
        sampler: &mut dyn Sampler,
    ) -> (Color, Vec3, Option<f32>) {
        let dir_out = -isect.ray.direction;
        let nrm_facing = isect.facing_normal;

        let mut transmit_threshold = 0.0;
        if !self.opaque {
            let reverse = isect.normal * dir_out < 0.0;

            let mut ratio = 1.0 / self.transmit_ior;
            if reverse {
                ratio = 1.0 / ratio;
            }

            let c1 = dir_out * nrm_facing;
            let c2 = (1.0 - ratio.powi(2) * (1.0 - c1.powi(2))).sqrt();

            let dir_in = Vec3::from(nrm_facing) * (ratio * c1 - c2) - dir_out * ratio;
            let throughput = self.transmitted(isect, -dir_out);
            transmit_threshold = throughput.max_component().min(1.0);
            let roulette = sampler.get_value();

            if roulette < transmit_threshold {
                let color =
                    self.transmitted(isect, dir_in) / (dir_out * nrm_facing * transmit_threshold);
                return (color, dir_in, None);
            }
        }

        let mut idx = 0;
        if self.brdfs.len() > 1 {
            let sample = sampler.get_value();
            idx = (((self.brdfs.len() as f32) * sample).floor() as usize).min(self.brdfs.len() - 1);
        }
        let brdf = &self.brdfs[idx];

        let dir_in = brdf.sample(sampler, nrm_facing, dir_out);
        let pdf = self.pdf(isect, dir_in);
        let color = self.reflected(isect, dir_in) / (1.0 - transmit_threshold);

        return (color, dir_in, Some(pdf));
    }

    pub fn pdf(&self, isect: &Intersection, dir_in: Vec3) -> f32 {
        let dir_out = -isect.ray.direction;
        let nrm_facing = isect.facing_normal;

        let mut total_pdf = 0.0;
        for brdf in self.brdfs.iter() {
            total_pdf += brdf.pdf(dir_in, nrm_facing, dir_out);
        }
        total_pdf /= self.brdfs.len() as f32;

        return total_pdf;
    }
}
